From f01ebd41487179fe61d6c0c6e5f75ff5ceabd21b Mon Sep 17 00:00:00 2001
From: Sarika Sharma <quic_sarishar@quicinc.com>
Date: Mon, 19 Aug 2024 11:58:39 +0530
Subject: [PATCH] wifi: ath12k: Add module param for 4addr NULL frame on
 WARN_ON_ONCE

Currently, WARN_ON_ONCE() is triggered if AP_VLAN is not created from
4addr null frame for WDS mix mode.

Add module param ap_vlan_without_4addr_null to avoid triggering
WARN_ON_ONCE() based on module param.

By default set ap_vlan_without_4addr_null to true to not trigger
WARN_ON_ONCE().

Patch-dependency: 920-wifi-mac80211-add-warn_on-for-sending-unexpected-
                  4addr-null-frame.patch

Signed-off-by: Sarika Sharma <quic_sarishar@quicinc.com>
---
 net/mac80211/mlme.c | 5 +++++
 net/mac80211/rx.c   | 4 +++-
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index f97b539..255a25e 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -71,6 +71,11 @@ module_param(debug_param, int, 0644);
 MODULE_PARM_DESC(debug_param,
 		"Debug module param to debug link id < 0");
 
+bool ap_vlan_without_4addr_null = true;
+module_param(ap_vlan_without_4addr_null, bool, 0644);
+MODULE_PARM_DESC(ap_vlan_without_4addr_null,
+		 "Check if AP_VLAN created without 4addr null frame");
+
 /*
  * Time the connection can be idle before we probe
  * it to see if we can still talk to the AP.
diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index 6d3584f..c032c17 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -34,6 +34,7 @@
 #include "rate.h"
 
 extern int debug_param;
+extern bool ap_vlan_without_4addr_null;
 #define IEEE80211_FSE_MAGIC_NUM      0xAA
 #define IEEE80211_FSE_MAGIC_NUM_MASK GENMASK(7,0)
 #define IEEE80211_PPE_VP_NUM         GENMASK(23, 8)
@@ -3366,7 +3367,8 @@ ieee80211_rx_h_data(struct ieee80211_rx_data *rx)
 			if (sdata->vif.offload_flags & IEEE80211_OFFLOAD_ENCAP_4ADDR) {
 				pr_warn("4addr non-null data frame: %d with frame_control: %x",
 					port_control, hdr->frame_control);
-				WARN_ON_ONCE(1);
+				if (!ap_vlan_without_4addr_null)
+					WARN_ON_ONCE(1);
 			}
 		}
 		if (!port_control)
-- 
2.34.1

